tee-worker/litentry/core/parachain-extrinsic-task/receiver/src/lib.rs

```diff
@@ -0,0 +1,53 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+use itp_extrinsics_factory::CreateExtrinsics;
+use itp_ocall_api::EnclaveOnChainOCallApi;
+use itp_types::parentchain::ParentchainId;
+use lc_parachain_extrinsic_task_sender::init_parachain_extrinsic_sender_storage;
+use log::*;
+use std::{format, string::String, sync::Arc, time, vec};
+
+pub fn run_parachain_extrinsic_task_receiver<ExtrinsicsFactory, OCallApi>(
+	api: Arc<OCallApi>,
+	extrinsic_factory: Arc<ExtrinsicsFactory>,
+) -> Result<(), String>
+where
+	ExtrinsicsFactory: CreateExtrinsics + Send + Sync + 'static,
+	OCallApi: EnclaveOnChainOCallApi,
+{
+	let task_receiver = init_parachain_extrinsic_sender_storage().map_err(|e| {
+		format!("Failed to initialize parachain extrinsic task sender storage: {:?}", e)
+	})?;
+	let mut calls = vec::Vec::new();
+
+	loop {
+		let start_time = time::Instant::now();
+		while start_time.elapsed() < time::Duration::from_secs(6) {
+			if let Ok(call) = task_receiver.recv() {
+				calls.push(call);
+			}
+		}
+		if !calls.is_empty() {
+			let extrinsic =
+				match extrinsic_factory.create_batch_extrinsic(calls.drain(..).collect(), None) {
+					Ok(extrinsic) => extrinsic,
+					Err(e) => {
+						error!("Failed to create extrinsic: {:?}", e);
+						continue
+					},
```

Comment:
If it errors out - does it mean the `calls` will be lost (as they are drained already)? There's no second chance for them to be sent.

Please also consider:
- `utility.batch` itself has a limit: `batched_calls_limit()`. It could be exceeded if we have a huge amount of `calls` within 6 sec
- the heavy (total) weight might constantly delay the parachain block production
- there should be a limit of byte array size going through the ocall

Please double-check those.
For the first two points an easy way is we do paged batch (like max 500 in one batch or so)

Reply:
We do have the problem of "huge amount of `calls` within 6 sec", based on the experience of last two campaign events.

Reply:
> If it errors out - does it mean the calls will be lost (as they are drained already)? There's no second chance for them to be sent.

I thought about this, if there is an error it means there was an issue with the cached data (metadata or nonce), so I'm not sure how we could recover from that, we could add some sort of retry mechanism if its possible to recover, but if its not, the data will just grow until it runs out of memory.

> Please also consider:
utility.batch itself has a limit: batched_calls_limit(). It could be exceeded if we have a huge amount of calls within 6 sec
the heavy (total) weight might constantly delay the parachain block production
there should be a limit of byte array size going through the ocall
Please double-check those.
For the first two points an easy way is we do paged batch (like max 500 in one batch or so)

Yes I checked this, the limit is almost 11k (It should be possible to get this value from the metadata), but I think it makes sense to create bounded batches to be safe 👍🏼 

> We do have the problem of "huge amount of calls within 6 sec", based on the experience of last two campaign events.

I think for our use case it should be fine as these extrinsics are only for emitting events

Reply:
> I thought about this, if there is an error it means there was an issue with the cached data (metadata or nonce), so I'm not sure how we could recover from that, we could add some sort of retry mechanism if its possible to recover, but if its not, the data will just grow until it runs out of memory.

Actually, if `send_to_parentchain` fails it makes sense to retry them, I'll update that

Reply:
After giving it some thought, I'm not sure a retry mechanism here would make sense unless we persist the calls somehow so it could retry again after restarting the worker for example. But I think this would be out of scope for this issue as the end result is not changing, currently if we have such errors, extrinsics are never sent

Reply:
When sending fails due to nonce clash (e.g. too low nonce, it could casually happen due to e.g. too late parachain sync), retrying will likely fix it.

You are using `utility.batch`, so it's possible that only some calls fail. If we apply a retrying mechanism, do we retry the whole batch or only those failed ext?

Reply:
The parachain ext could mutate the parachain state, so it seems we would need them to be sent exactly once, not more, not less

Reply:
> You are using utility.batch, so it's possible that only some calls fail. If we apply a retrying mechanism, do we retry the whole batch or only those failed ext?

This is very good point, I think it would be a bit hard to retry the remaining calls. As soon as one of the calls in the batch fails an event is emitted (`BatchInterrupted`) with the index of the call and the error, then the batch call stops there, so we would have to listen to those events somehow and send another batch with the remaining calls from that index. 

Another option would be to use `batch_all` which is an atomic operation, so the transaction will rollback and fail if any of the calls failed

Reply:
> When sending fails due to nonce clash (e.g. too low nonce, it could casually happen due to e.g. too late parachain sync), retrying will likely fix it.

Yes, you are right. The issue is that with the current implementation there is no way to know that the call failed because of invalid nonce, we get an error only if the extrinsic could not be decoded in which case I don't think there is something we can do to recover.

Reply:
I think we'll need to have SendingStrategy here again, substrate-api-client already has some useful state indication, like if waiting for inclusion or not, if waiting for success or not.

But that's out of this PR's scope